keps/sig-windows/20190714-windows-csi-support.md

```diff
@@ -302,14 +303,12 @@ Once compiled for Windows, container images based on Window Base images (like Na
 
-![Kubernetes CSI Node Components and Interactions](https://ddebroywin1.s3-us-west-2.amazonaws.com/csi-proxy2.png "Kubernetes CSI Node Components and Interactions")
+![Kubernetes CSI Node Components and Interactions](csi-proxy3.png?raw=true "Kubernetes CSI Node Components and Interactions")
 
 A GRPC based interface, CSIProxyService, will be used by CSI Node Plugins to invoke privileged operations on the host through csi-proxy.exe. CSIProxyService will be versioned and any release of csi-proxy.exe binary will strive to maintain backward compatibility across as many prior stable versions of the API as possible. This avoids having to run multiple versions of the csi-proxy.exe binary on the same Windows host if multiple CSI Node Plugins (that do not depend on APIs in Alpha or Beta versions) have been configured and the version of the csi-proxy API required by the plugins are different. For every version of the API supported, csi-proxy.exe will first probe for and then expose a `\\.\pipe\csi-proxy-v[N]` pipe where v[N] can be v1, v2alpha1, v3beta1, etc. If during the initial probe phase, csi-proxy.exe determines that another process is already listening on a `\\.\pipe\csi-proxy-v[N]` named pipe, it will not try to create and listen on that named pipe. This allows multiple versions of csi-proxy.exe to run side-by-side if absolutely necessary to support multiple CSI Node Plugins that require widely different versions of CSIProxyService that no single version of csi-proxy.exe can support.
 
-If an old stable version of CSIProxyService, say vN, can no longer be supported, support for it may be dropped from the latest version of csi-proxy.exe. To continue running CSI Node Plugins that depend on the old version vN, Kubernetes administrators will be required to run the latest version of the csi-proxy.exe (that will be used by CSI Node Plugins that use versions of CSIProxyService more recent than vN) along with an old version of csi-proxy.exe that does support vN.
-
 The following are the main RPC calls that will comprise a v1 version of the CSIProxyService API:
 
 ```
```

Comment:
are the RPC messages described below (MkdirRequest, RmdirRequest, etc) pre-existing or introduced as part of this feature?

since this is exposing an API that allows things running inside containers to escalate permissions to do host-level operations, we need clear information about the following aspects:
* what values are allowed in paths
  * are special characters restricted?
  * are backstep path segments allowed?
  * are forward and backslashes (and international path separators like ¥) allowed and normalized?
  * which of the filepath variants in https://docs.microsoft.com/en-us/dotnet/standard/io/file-path-formats are allowed?
* is the csi-proxy process or the caller responsible for normalization and validation of paths?
* do symlink traversal and time-of-check-time-of-use issues exist in windows? for example, could a container create a link in a filesystem it controlled that pointed to a host filepath it should not have access to, then request a CSI volume be mounted at a path beneath that link, and get the privileged csi-proxy to follow the link and overwrite/expose host files to the container? if so, how does the caller indicate to the csi-proxy what the acceptable boundary is for resolved paths that include links?

Reply:
It's possible some or many of the validation bits are constrained by the CSI spec already and the kubelet is responsible for filtering out some of those things. If so, just noting that and ensuring test coverage of the kubelet is fine. 

The symlink/link/traversal issues are more interesting, since the csi-proxy would be the only actor capable of performing those checks.

Reply:
All the RPCs mentioned are new. Will clarify the path semantics and validations.

Reply:
Rather than repeating everything in the doc - I think we can reference https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file

@liggitt - that doc you linked to is a subset implemented in .Net. https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getfullpathnamew is the native API that should be used to canonicalize file names that follows the rules listed in the prior doc

Reply:
+1 for linking to existing references, clarifying which API is used, and being explicit about where normalization happens (which component(s) are responsible for doing it)

Reply:
path normalization on the node side can be done with https://github.com/golang/go/blob/2b8cbc384d092dc63e9dc18ec318d0682185611c/src/syscall/syscall_windows.go#L224

Reply:
I have updated the Request and Response struct descriptions for the RPCs. /cc @liggitt and @PatrickLang PTAL.

Reply:
Thanks. Do these map closely to windows filesystem APIs? If so, that would be good to reference.

I'll defer to the windows/storage experts to determine if these provide a reasonable API that covers the information/operations a CSI driver would need.